src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp

```diff
@@ -0,0 +1,125 @@
+#include "ParallelAlgorithms/Interpolation/Protocols/PostInterpolationCallback.hpp"
+#include "Utilities/Functional.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/PrettyType.hpp"
+#include "Utilities/ProtocolHelpers.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+namespace db {
+template <typename TagsList>
+class DataBox;
+}  // namespace db
+namespace observers {
+template <class Metavariables>
+struct ObserverWriter;
+}  // namespace observers
+/// \endcond
+
+namespace intrp {
+namespace callbacks {
+
+/// \brief post_interpolation_callback that outputs
+/// 2D "volume" data on a surface.
+///
+/// Uses:
+/// - Metavariables
+///   - `temporal_id`
+/// - DataBox:
+///   - `TagsToObserve` (each tag must be a Scalar<DataVector>)
+///
+/// Conforms to the intrp::protocols::PostInterpolationCallback protocol
+///
+/// For requirements on InterpolationTargetTag, see
+/// intrp::protocols::InterpolationTargetTag
+template <typename TagsToObserve, typename InterpolationTargetTag>
+struct ObserveSurfaceData
+    : tt::ConformsTo<intrp::protocols::PostInterpolationCallback> {
+  static constexpr double fill_invalid_points_with =
+      std::numeric_limits<double>::quiet_NaN();
+
+  using const_global_cache_tags = tmpl::list<observers::Tags::SurfaceFileName>;
+
+  template <typename DbTags, typename Metavariables, typename TemporalId>
+  static void apply(const db::DataBox<DbTags>& box,
+                    Parallel::GlobalCache<Metavariables>& cache,
+                    const TemporalId& temporal_id) {
+    const Strahlkorper<Frame::Inertial>& strahlkorper =
+        get<StrahlkorperTags::Strahlkorper<Frame::Inertial>>(box);
+    const YlmSpherepack& ylm = strahlkorper.ylm_spherepack();
+    const std::array<DataVector, 2>& theta_phi = ylm.theta_phi_points();
+    const DataVector& theta = theta_phi[0];
+    const DataVector& phi = theta_phi[1];
+    const DataVector& sin_theta = sin(theta);
+    const DataVector& radius = ylm.spec_to_phys(strahlkorper.coefficients());
+
+    // Output the inertial-frame coordinates.
+    const std::string& surface_name =
+        pretty_type::name<InterpolationTargetTag>();
+    std::vector<TensorComponent> tensor_components{
+        {surface_name + "/InertialCoordinates_x"s,
+         radius * sin_theta * cos(phi)},
+        {surface_name + "/InertialCoordinates_y"s,
+         radius * sin_theta * sin(phi)},
+        {surface_name + "/InertialCoordinates_z"s, radius * cos(theta)}};
+
+    // Output each tag if it is a scalar. Otherwise, throw a compile-time error.
+    // This could be generalized to loop over components, outputting them one at
+    // a time.
+    tmpl::for_each<TagsToObserve>([&box, &surface_name,
+                                   &tensor_components](auto tag_v) {
+      using Tag = tmpl::type_from<decltype(tag_v)>;
+      static_assert(std::is_same_v<typename Tag::type, Scalar<DataVector>>,
+                    "Each tag in TagsToObserve must be a Scalar<DataVector>.");
```

Comment:
It might be helpful to give some justification why, and/or how/if this can be extended. This should also be noted in the documentation

Reply:
The generalization would be to output non-scalar components one at a time, treating each as a scalar. But in practice this isn't necessary, because we've never visualized vectors or higher-ranked tensors on a Strahlkorper, and I'm not sure how useful that would even be. I'll clarify the comment at line 89 along these lines.

Reply:
Sorry, I was thinking of expanding the `static_assert` message to include the reason why. That way people don't need to find the source code and hope there's a comment. I'm fine with having a longer C++ comment with a shorter `This could be generalized if needed, see the code comment above the assert.`